Show and hide functionality (simple state preserving).

jonholmberg opened the pull request:
Added a show/hide button that shows/hides a fragment containing the message sent. Removed the `DisplayMessageActivity`. Also, I tried to move more functionality into the extended `Fragment` classes instead of leaving it in the activity.

The state of the runtime added fragment should now be preserved, so that the button shows the correct text, the visibility of the fragment is preserved and the message is preserved during configuration changes. The `MainActivity` preserves the visibility state of the `DisplayMessageFragment` and the `DisplayMessageFragment` preserves the message.

Oh, and don't mind all the commits, most of them are just me testing stuff.

jonholmberg committed:
The DisplayMessageFragment can now be added by the user on runtime in landscape

jonholmberg committed:
Added Clear messages button (not working as intended in DisplayMessageActivity)

jonholmberg committed:
The fragment that shows the message now pops up dynamically

jonholmberg committed:
Added some test retaining functionality

jonholmberg committed:
The text on the show/hide button now works as intented

jonholmberg committed:
Added DisplayMessageFragment controller value in MainActivity

jonholmberg committed:
Hide and show function now works as intended

jonholmberg committed:
Cleaned up and commented

jonholmberg commented on jlh/JonFirstApp/app/src/main/java/com/e_conomic/jonfirstapp/MainActivity.java:
```diff
@@ -1,68 +1,146 @@
+    }
+
+    /** Helper function to change the layout of the fragment container */
+    private void changeFragmentLayout(View fragment_layout, int newLayoutParam) {
```
Please tell me there's another way to hide and show a fragment (**and** dynamically change the view of the fragment besides) than what implemented in this method? The hide and show functions I used doesn't quite seem to do the trick. For instance, when on landscape, it leaves the entire right side white, instead of "stretching" the main fragment to fill the entire screen 😨

AndersHqst replied:
@tudorgk what is your take on this?

jonholmberg commented on jlh/JonFirstApp/app/src/main/java/com/e_conomic/jonfirstapp/MainActivity.java:
```diff
@@ -1,68 +1,146 @@
+            isDisplayMessageFragmentVisible = false;
+
+            // Inform the main fragment of the change
+            main_fragment.setButtonText();
```
Is there a nice way to create a listener on the `isDisplayMessageFragmentVisible` so that I don't have to call `setButtonText()` when I change the value? They show a way [here](https://developer.android.com/training/basics/fragments/communicating.html), but it seems a bit overkill here?

tudorgk replied:
IMO this is before we had libraries like [`EventBus`](https://github.com/greenrobot/EventBus). Try looking into that.

AndersHqst replied:
@jonholmberg the design you link to is the way to go, and a standard pattern for this particular situation, that is also used everywhere on iOS. Currently your fragment depends on a `MainActivity` class. It is much better to design your fragment to communicate though an interface that defines the `setButtonText` method. That way, the fragment is decoupled and easy to reuse with other activities that implement the interface. It also makes it easier to understand how your fragment works, because its described by the interface. Let's draw this on a board next time to make sure the design is clear.

jonholmberg commented on jlh/JonFirstApp/app/src/main/java/com/e_conomic/jonfirstapp/DisplayMessageFragment.java:
```diff
@@ -1,17 +1,49 @@
 public class DisplayMessageFragment extends Fragment {
 
+    // The message to be displayed.
+    private String message = "";
```
Should this be in resources instead?

tudorgk replied:
Wouldn't recommend it, resources should hold only static values that don't change over time.

jonholmberg replied:
Cool 👍

tudorgk commented on jlh/JonFirstApp/app/src/main/res/layout/fragment_main.xml:
```diff
@@ -50,4 +50,11 @@
+        android:layout_width="match_parent"
+        android:layout_height="wrap_content"
+        android:text="@string/button_showmsg"
+        android:onClick="showHide"/>
```
I think you should probably want to avoid writing onClick listeners in the xml, in the future. For your sanity's sake anyway...

tudorgk commented on jlh/JonFirstApp/app/src/main/java/com/e_conomic/jonfirstapp/DisplayMessageFragment.java:
```diff
@@ -1,17 +1,49 @@
+
+    /** Updates the TextView that shows the message. */
+    public void updateMessage(String newMessage) {
+        message = newMessage;
```
Do we need to keep track of the `message` variable here? What I mean by this is that we don't use the `message` variable anywhere else and we can set the `newMessage` directly to the TextView. If you would implement the `onSaveInstanceState` method, it would become useful, otherwise I would just remove it.

jonholmberg replied:
Hmm.. but I also use it in `onViewCreated` to set the message after a configuration change? :) E.g. if you turn your phone it will still show the message you entered before you turned it!

tudorgk commented:
Please don't forget to rename the layout files, because it makes it difficult to know which layout belongs to which Fragment class.

tudorgk commented on jlh/JonFirstApp/app/src/main/java/com/e_conomic/jonfirstapp/MainActivity.java:
```diff
@@ -1,68 +1,146 @@
+
+        if (isDisplayMessageFragmentVisible) {
+            // Set width of fragment layout container to fill half the screen.
+            changeFragmentLayout(findViewById(R.id.fragment_container),
```
So the fragment and the container have the same id `fragment_container`? I'm a little bit confused :)

jonholmberg replied:
Oops.. changed it :)

tudorgk commented on jlh/JonFirstApp/app/src/main/java/com/e_conomic/jonfirstapp/MainActivity.java:
```diff
@@ -1,68 +1,146 @@
-        // The message entered by the user.
-        String message = editText.getText().toString();
+        // If the display message fragment is visible show the message there.
+        if (display_message_fragment != null) {
```
Maybe you would want to log this part. We should be able to know why or when is the fragment null.

AndersHqst replied:
I commented on this above as well. Will a log solve/avoid the issue?

tudorgk commented on jlh/JonFirstApp/app/src/main/java/com/e_conomic/jonfirstapp/MainActivity.java:
```diff
@@ -1,68 +1,146 @@
         setContentView(R.layout.fragment_layout);
-        editText = (EditText) findViewById(R.id.edit_message);
+
+        FragmentManager fragmentManager = getSupportFragmentManager();
```
I would extract the remaining lines into their own method. Makes understanding the code easier and give out cleaner code.

jonholmberg replied:
Good idea.

tudorgk commented on jlh/JonFirstApp/app/src/main/java/com/e_conomic/jonfirstapp/MainActivity.java:
```diff
@@ -1,68 +1,146 @@
-            // Create intent.
-            Intent intent = new Intent(this, DisplayMessageActivity.class);
+        // Setup FragmentManager and FragmentTransaction.
+        View fragment_container = findViewById(R.id.fragment_container);
```
Here, the `fragment_container` should be cached somewhere else. We don't want to do a layout pass every time we want to show or hide the message.

jonholmberg replied:
Performance wise, would you say it's better to get the `fragment_container` when the `DisplayMessageFragment` is created in `onCreate` (even though you actually may never press the show/hide button) or make a check here to see if it has already been found? Then you have to do the check every time you press the button. I mean, it's a small matter, but I'm just curious :)

tudorgk commented on jlh/JonFirstApp/app/src/main/java/com/e_conomic/jonfirstapp/MainActivity.java:
```diff
@@ -1,68 +1,146 @@
+
+        // Hide Fragment if it is visible.
+        if (display_message_fragment.isVisible()) {
+            ft.hide(display_message_fragment).commitNow();
```
I would consider saving the fragment instance in the `MainActivity` and instead of showing/hiding the fragment, maybe it would be better to `add` and `remove` it from the view hierarchy. By doing so, I think we can remove the necessity of maintaining `changeFragmentLayout` logic. Check [this link](https://developer.android.com/training/basics/fragments/fragment-ui.html) for inspiration. (Instead of replacing we can just remove/add the fragment). Haven't tested this myself, maybe it's worth a try.

jonholmberg replied:
I will give it a try!

tudorgk commented on jlh/JonFirstApp/app/src/main/java/com/e_conomic/jonfirstapp/MainActivity.java:
```diff
@@ -1,68 +1,146 @@
+    }
+
+    /** Helper function to change the layout of the fragment container */
+    private void changeFragmentLayout(View fragment_layout, int newLayoutParam) {
```
Parameter names should be generic and should have the same style. You have to choose either `camelCase` or `snake_case`. We usually use `camelCase`. `snake_case` is usually seen in low level C programs. Also the `newLayoutParam` type is ambiguous, I would change the method to support an actual `LayoutParam` class. For example instead of:
```kotlin
private void changeFragmentLayout(View fragment_layout, int newLayoutParam)
```
it should be:
```kotlin
private void changeFragmentLayout(View viewGroup, ViewGroup.LayoutParams newLayoutParam)
```

jonholmberg replied:
Fair point!

tudorgk commented on jlh/JonFirstApp/app/src/main/java/com/e_conomic/jonfirstapp/MainActivity.java:
```diff
@@ -1,68 +1,146 @@
-            startActivity(intent);
+            // Set width of fragment layout container to fill half the screen.
+            changeFragmentLayout(fragment_container, ViewGroup.LayoutParams.MATCH_PARENT);
+            return;
```
`return` is not necessary because the method returns `void`

tudorgk commented on jlh/JonFirstApp/app/src/main/java/com/e_conomic/jonfirstapp/MainActivity.java:
```diff
@@ -1,68 +1,146 @@
+        }
+
+        // Show Fragment if it is hidden.
+        if (display_message_fragment.isHidden()) {
```
The control flow here should be
```java
if (display_message_fragment.isVisible()) {
  // ...
} else {
  // ...
}
```

jonholmberg replied:
I thought I read somewhere that a fragment could be neither visible nor hidden, but that doesn't seem quite right...?

tudorgk commented on jlh/JonFirstApp/app/src/main/java/com/e_conomic/jonfirstapp/MainActivity.java:
```diff
@@ -1,68 +1,146 @@
+
+            // Set width of fragment layout container to 0.
+            changeFragmentLayout(fragment_container, 0);
+            return;
```
`return` is not necessary because the method returns `void`

tudorgk commented on jlh/JonFirstApp/app/src/main/res/layout-land/fragment_layout.xml:
```diff
@@ -4,17 +4,17 @@
-        android:layout_height="match_parent">
+    <FrameLayout
+        android:id="@+id/fragment_container"
+        android:layout_width="0sp"
```
`sp` should be used mostly for text. When setting view heights/widths we should use `dp`. Please check out [this link](http://stackoverflow.com/a/2025541).

jonholmberg replied:
Ah right, thanks!

tudorgk commented on jlh/JonFirstApp/app/src/main/res/layout/display_message.xml:
```diff
@@ -1,14 +1,18 @@
 <?xml version="1.0" encoding="utf-8"?>
 <LinearLayout xmlns:android="http://schemas.android.com/apk/res/android"
-    android:id="@+id/activity_display_message"
+    android:id="@+id/display_message"
```
Check naming.

AndersHqst replied:
@tudorgk could you elaborate

jonholmberg replied:
I get it. There's no longer an activity for this, only a fragment.

tudorgk reviewed (changes requested):
First of all, I want to say that you did some very nice work!! 🙌 
Don't worry about the number of comments. We have a learning process going on here. 😄

AndersHqst commented:
@jonholmberg try and be consistent in how you format and name your branches. I see `foo_bar`, `foo-bar`, and `fooBar`. Pick your religion and stick with it. When working on our projects it will be `foo-bar`. Also, we normally try to name branches verbosely e.g. `show-and-hide-functionality` - I honestly said "Show and hide functor" in my head when I read your branch name :]

AndersHqst commented:
Regarding "all the commits", we'll take a look at how you can clean up your commit history.

AndersHqst commented on jlh/JonFirstApp/app/src/main/java/com/e_conomic/jonfirstapp/DisplayMessageFragment.java:
```diff
@@ -1,17 +1,49 @@
+    private TextView display_message;
+
+    @Override
+    public void onCreate(@Nullable Bundle savedInstanceState) {
```
Do you understand `@Nullable` ?

jonholmberg replied:
I think so. It indicates that the parameter could be `null` so that you have to handle that case in the method body. Here, it's handled by the `super.onCreate(..)`. I guess it's like the optional parameters you have in Kotlin? :)

AndersHqst replied:
Almost, it is true that it indicates that the parameter could be null, but you won't be forced to handle it. Only thing to say about `super.onCreate(..)` is that that should ideally also declare the parameter as `@Nullable` There are tools like "FindBugs" that can use this annotation to give you static warnings if your do not check for null. Maybe the IDE is clever enough to also give you some help too. There is also `@NotNull`.

In Kotlin, however, you will be forced to check for null. Your code simply won't be able to compile if you don't. This it is not done with an annotation like here, but built into the type system as "Optionals". For example, here it would be an "Optional Bundle", and you will be forced to check if the "Optional Bundle" is a "Bundle" or "null".

AndersHqst commented on jlh/JonFirstApp/app/src/main/java/com/e_conomic/jonfirstapp/DisplayMessageFragment.java:
```diff
@@ -1,17 +1,49 @@
+    // The TextView that displays the message.
+    private TextView display_message;
+
+    @Override
```
Do you know what `@Override` does?

jonholmberg replied:
It lets the compiler and IDE  know that the following method should override a method in the superclass? :)

AndersHqst replied:
Exactly, avoids issues where you might misspell a method name. In Kotlin, this kind of thing will be enforced.

AndersHqst commented on jlh/JonFirstApp/app/src/main/java/com/e_conomic/jonfirstapp/DisplayMessageFragment.java:
```diff
@@ -1,17 +1,49 @@
+    /** Updates the TextView that shows the message. */
+    public void updateMessage(String newMessage) {
+        message = newMessage;
+        if (display_message != null) {
```
This is fine guard, but as an exercise, consider this: why/when would `display_message` be null? And if it is null, then what? Our app will just stop showing the message and seem broken to the user.

jonholmberg replied:
Fair point. I made sure it is initialised in `onViewCreated(..)` so that when the fragment is created it should not be `null`! So the reason for the check was actually if the garbage collector somehow removed the view or the reference to the view?. I guess if it wasn't used for some time.

I'm not sure if either can be the case, but then you would have to either recreate the view or update the view reference.

Edit: I'm actually not quite sure how this works, but if the view was removed, would the `display_message` still just be a reference to the address where the view used to be, or would it be `null`?

AndersHqst commented on jlh/JonFirstApp/app/src/main/java/com/e_conomic/jonfirstapp/MainActivity.java:
```diff
@@ -1,68 +1,146 @@
+        fragment_layout.setLayoutParams(layoutParams);
+    }
 
+    /** Retrieves the isDisplayMessageFragmentVisible boolean. */
```
Consider whether this is good documentation. I see you add several comments that function as self-notes for you to learn and remember - which is great for now. Moving forward, think of comments as as way to explain code that is not self-explanatory and to document. For example, for this method you would want some documentation, so that when using the method it is described in plain English what `isDisplayMessageFragmentVisible` is. Currently your method documentation here says nothing more than what the method signature already says.

jonholmberg replied:
Good point!

AndersHqst commented on jlh/JonFirstApp/app/src/main/java/com/e_conomic/jonfirstapp/MainFragment.java:
```diff
@@ -1,18 +1,71 @@
 
 public class MainFragment extends Fragment {
 
+    // The main activity.
```
Delete comment, code is self-explanatory

AndersHqst commented on jlh/JonFirstApp/app/src/main/java/com/e_conomic/jonfirstapp/MainFragment.java:
```diff
@@ -1,18 +1,71 @@
 public class MainFragment extends Fragment {
 
+    // The main activity.
+    MainActivity main_activity;
```
Name formatting on Android seems to be a bit confusing at times. Here, I would say it should be `mainActivity`

jonholmberg replied:
What about the fragment references? Say, in `MainActivity.java` should I also write `displayMessageFragment` instead of `display_message_fragment`? :)

AndersHqst commented on jlh/JonFirstApp/app/src/main/java/com/e_conomic/jonfirstapp/MainFragment.java:
```diff
@@ -1,18 +1,71 @@
+        try {
+            main_activity = (MainActivity) context;
+        } catch (ClassCastException e) {
+            throw new ClassCastException(context.toString() + " is not an activity.");
```
It might be an activity, but certainly not of type `MainActivity`

AndersHqst commented:
Great work @jonholmberg. Take your time to work through the review and don't hesitate asking us questions back.

jonholmberg committed:
Cleaned up a bit and renamed files and id's to be more consistent

jonholmberg committed:
Lunch commit

jonholmberg committed:
Cleaning up: renaming and removing redundancies

jonholmberg committed:
Cleaning up: renaming and removing redundancies

AndersHqst commented on jlh/JonFirstApp/app/src/main/java/com/e_conomic/jonfirstapp/MainActivity.java:
```diff
@@ -139,7 +139,7 @@ private void changeFragmentLayout(View fragment_layout, int newLayoutParam) {
     }
 
-    /** Retrieves the isDisplayMessageFragmentVisible boolean. */
+    /** Check whether the fragment that displays the message is visible. */
```
Better. Common jargon could be something like "Returns true if this fragment is visible"

AndersHqst commented on jlh/JonFirstApp/app/src/main/java/com/e_conomic/jonfirstapp/MainFragment.java:
```diff
@@ -13,8 +12,7 @@
 
-    // The main activity.
-    MainActivity main_activity;
+    MainActivity mainActivity;
```
Thought I saw you asking a question about the naming here somewhere. The ~convention for camelCase vs underscore_case in Android (in our team) is that ids and names constructed for us on the fly are underscore_case, otherwise camelCase

AndersHqst commented on jlh/JonFirstApp/app/src/main/java/com/e_conomic/jonfirstapp/MainFragment.java:
```diff
@@ -25,9 +23,9 @@ public void onAttach(Context context) {
         // Retrieve the main activity.
         try {
-            main_activity = (MainActivity) context;
+            mainActivity = (MainActivity) context;
```
A take away here is that a `Fragment` is born with a reference to the owning `Activity`. We will use this when we look at the fragment to activity, or other fragment, callback pattern we discussed earlier.

AndersHqst commented on jlh/JonFirstApp/app/src/main/java/com/e_conomic/jonfirstapp/DisplayMessageFragment.java:
```diff
@@ -43,6 +43,8 @@ public void onViewCreated(View view, @Nullable Bundle savedInstanceState) {
     public void updateMessage(String newMessage) {
         message = newMessage;
+
+        // TODO: what if it is null?
```
Comments are gone. But anyway, my point was this. If the `display_message`, for whatever weird reason (maybe some other developer came around and accidentally renamed an id) , is null, your app will seem broken. So what to do? One way would be to not check for null and let the app crash. I often consider that a better solution, because then the app will at least be restarted and then maybe work, and we are not leaving the user in doubt that things didn't work. As you will see in Kotlin with Optionals, we will often _force unwrap_ this kind of thing to say "this thing is nullable, but blow up if it is"

I would say a great solution is to force unwrap (Kotlin), and having a __test__ that checks that this it not null, whenever the fragment has been created. Others may feel differently.

AndersHqst commented on jlh/JonFirstApp/app/src/main/java/com/e_conomic/jonfirstapp/MainActivity.java:
```diff
@@ -60,7 +62,7 @@ private void setupFragments() {
 
         // Setup main fragment
-        main_fragment = (MainFragment) fragmentManager.findFragmentById(R.id.fragment_main);
+        mainFragment = (MainFragment) fragmentManager.findFragmentById(R.id.fragment_main);
```
This looks nice. At a later point you will see that the `findFragmentById` dance can be avoided, in which case an auto-generated `main_fragment` will be magically available.

AndersHqst commented:
Great work @jonholmberg. I saw a few other comments from you that I can't seems to find now. Please review the rest of our comments, and let us discuss this tomorrow. Then I think we are ready to merge :]

jonholmberg committed:
Merge branch 'showAndHideFunct' of github.com:e-conomic/android-course into showAndHideFunct

jonholmberg committed:
Added MainFragmentListener interface and created onClickListeners in java code

jonholmberg committed:
Added MainFragmentListener interface and created onClickListeners in java code

jonholmberg committed:
Merge branch 'showAndHideFunct' of github.com:e-conomic/android-course into showAndHideFunct

jonholmberg committed:
Cleaned up the code that changes the layout to make it more readable

jonholmberg commented on jlh/JonFirstApp/app/src/main/java/com/e_conomic/jonfirstapp/MainActivity.java:
```diff
@@ -1,68 +1,160 @@
+     *
+     * @param viewGroup The view that receives a layout change.
+     * @param newLayoutParams The new layout parameters that will be added to the given view. */
+    private void changeFragmentLayout(View viewGroup, ViewGroup.LayoutParams newLayoutParams) {
```
Is this better now @tudorgk ?

jonholmberg commented:
Note that I moved the `isDisplayMessageFragmentVisible` boolean to `MainFragment.java` (used to be in `MainActivity.java`) as it resulted in much less communication back and forth between the fragment and activity :)

jonholmberg committed:
Removed changeFragmentLayout method

Button reference fix

Button reference fix

Random generated cange

Removed changeFragmentLayout method

tudorgk commented on jlh/JonFirstApp/app/src/main/java/com/e_conomic/jonfirstapp/MainActivity.java:
```diff
@@ -1,68 +1,160 @@
+     *
+     * @param viewGroup The view that receives a layout change.
+     * @param newLayoutParams The new layout parameters that will be added to the given view. */
+    private void changeFragmentLayout(View viewGroup, ViewGroup.LayoutParams newLayoutParams) {
```
I don't think we need this method anymore. What we want now is to set the layout params from the mapping. So we should just call:
```java
displayMessageFragmentContainer.setLayoutParams(fragmentLayouts.get(PORTRAIT_HIDE_FRAGMENT_LAYOUT));
```

jonholmberg replied:
Ah true! I will change it :)

tudorgk commented on jlh/JonFirstApp/app/src/main/java/com/e_conomic/jonfirstapp/MainActivity.java:
```diff
@@ -1,68 +1,160 @@
-            startActivity(intent);
+            // Set main fragment to fill entire screen.
+            if (displayOrientation == LANDSCAPE) {
+                changeFragmentLayout(displayMessageFragmentContainer,
```
Method not needed anymore.

tudorgk commented on jlh/JonFirstApp/app/src/main/java/com/e_conomic/jonfirstapp/MainActivity.java:
```diff
@@ -1,68 +1,160 @@
+                changeFragmentLayout(displayMessageFragmentContainer,
+                        fragmentLayouts.get(LANDSCAPE_HIDE_FRAGMENT_LAYOUT));
+            } else {
+                changeFragmentLayout(displayMessageFragmentContainer,
```
Method not needed anymore.

tudorgk commented on jlh/JonFirstApp/app/src/main/java/com/e_conomic/jonfirstapp/MainActivity.java:
```diff
@@ -1,68 +1,160 @@
+            ft.show(displayMessageFragment).commitNow();
+
+            // Set size of displayMessageFragment layout container to fill half the screen.
+            changeFragmentLayout(displayMessageFragmentContainer,
```
Method not needed anymore.

tudorgk reviewed (commented):
These are my last comments, I promise😅 . When you are ready, you can merge it.

AndersHqst commented:
Great work guys!

jonholmberg commented:
(poolparty)
